contracts/src/registry/EnhancedAccessControl.sol

```diff
@@ -0,0 +1,220 @@
+
+    /**
+     * @dev user role within a context.
+     * Context -> Role -> User -> RoleAssignment
+     */
+    mapping(bytes32 context => mapping(bytes32 role => mapping(address account => RoleAssignment roleAssignment))) private _roles;
+
+    /**
+     * @dev We use a version number to track the changes to the role assignments.
+     * This means we can easily remove all assignments of a given role without needing a loop.
+     */
+    mapping(bytes32 context => mapping(bytes32 role => uint256 version)) private _roleVersion;
+
+    /**
+     * @dev The root context.
+     */
+    bytes32 public constant ROOT_CONTEXT = bytes32(0);
+
+    /**
+     * @dev The default admin role.
+     */
+    bytes32 public constant DEFAULT_ADMIN_ROLE = 0x00;
+
+    /**
+     * @dev Modifier that checks that sender has a specific role within the given context. 
+     * Reverts with an {AccessControlUnauthorizedAccount} error including the required role.
+     */
+    modifier onlyRole(bytes32 context, bytes32 role) {
+        _checkRole(context, role);
+        _;
+    }
+
+    /**
+     * @dev See {IERC165-supportsInterface}.
+     */
+    function supportsInterface(bytes4 interfaceId) public view virtual override returns (bool) {
+        return interfaceId == type(EnhancedAccessControl).interfaceId || super.supportsInterface(interfaceId);
+    }
+
+    /**
+     * @dev Returns `true` if `account` has been granted `role`.
+     */
+    function hasRole(bytes32 context, bytes32 role, address account) public view virtual returns (bool) {
+        uint256 roleVersion = _roleVersion[context][role];
+        uint256 roleVersionInRoot = _roleVersion[ROOT_CONTEXT][role];
+
+        return (_roles[context][role][account].hasRole && _roles[context][role][account].version == roleVersion) ||
+               (_roles[ROOT_CONTEXT][role][account].hasRole && _roles[ROOT_CONTEXT][role][account].version == roleVersionInRoot);
+    }
+
+    /**
+     * @dev Reverts with an {AccessControlUnauthorizedAccount} error if `_msgSender()`
+     * is missing `role`. Overriding this function changes the behavior of the {onlyRole} modifier.
+     */
+    function _checkRole(bytes32 context, bytes32 role) internal view virtual {
+        _checkRole(context, role, _msgSender());
+    }
+
+    /**
+     * @dev Reverts with an {AccessControlUnauthorizedAccount} error if `account`
+     * is missing `role`.
+     */
+    function _checkRole(bytes32 context, bytes32 role, address account) internal view virtual {
+        if (!hasRole(context, role, account)) {
+            revert EnhancedAccessControlUnauthorizedAccount(context, role, account);
+        }
+    }
+
+    /**
+     * @dev Returns the admin role that controls `role`. See {grantRole} and
+     * {revokeRole}.
+     *
+     * To change a role's admin, use {_setRoleAdmin}.
+     */
+    function getRoleAdmin(bytes32 role) public view virtual returns (bytes32) {
+        return _adminRoles[role];
```

Comment:
Shouldn't this have a fallthrough case for the default admin role?

Reply:
Default admin role is 0x00 so fallthrough is automatic.